File: backend/schemas/review.py
from typing import Optional

from pydantic import BaseModel, Field


class AnalyzeRequest(BaseModel):
    owner: str = Field(..., description="仓库所有者")
    repo: str = Field(..., description="仓库名称")
    pr_number: int = Field(..., description="PR 编号")
    post_comment: bool = Field(False, description="分析完成后是否自动发布评论到 PR")


class FileInfo(BaseModel):
    filename: str
    status: str
    additions: int
    deletions: int
    patch: str


class PRInfoResponse(BaseModel):
    owner: str
    repo: str
    number: int
    title: str
    description: str
    author: str
    base_branch: str
    head_branch: str
    files_changed: int
    additions: int
    deletions: int
    files: list[FileInfo]
    diff_content: str


class RiskItem(BaseModel):
    severity: str
    file: str
    line: int
    description: str
    suggestion: str
    confidence: Optional[float] = None
    is_false_positive: bool = False


class Suggestion(BaseModel):
    category: str
    description: str
    file: str
    code_snippet: str = ""
    confidence: Optional[float] = None


class Discrepancy(BaseModel):
    type: str = Field(..., description="不一致类型: scope_drift | hidden_breaking | missing_desc | unrelated_file")
    description: str
    file: str = ""
    severity: str = "medium"


class IntentCheck(BaseModel):
    declared_intent: str = ""
    actual_scope: str = ""
    consistency_score: int = Field(100, description="意图一致性评分 0-100, 100=完全一致")
    verdict: str = Field("match", description="判定: match | minor_deviation | major_deviation")
    discrepancies: list[Discrepancy] = []


class AnalyzeResponse(BaseModel):
    pr_info: PRInfoResponse
    summary: str = ""
    risk_items: list[RiskItem] = []
    suggestions: list[Suggestion] = []
    risk_score: int = 0
    risk_level: str = "low"
    estimated_review_minutes: int = 0
    context_coverage: Optional[float] = None
    rule_matches: list["RuleMatch"] = []
    analysis_id: Optional[int] = None
    intent_check: Optional[IntentCheck] = None
    risk_clusters: list["RiskClusterItem"] = []


class RuleMatch(BaseModel):
    rule_id: int
    rule_name: str
    severity: str
    file: str
    line: int
    matched_text: str
    suggestion: Optional[str] = None


class RiskClusterItem(BaseModel):
    category: str
    label: str
    risk_indices: list[int] = []
    dominant_severity: str = "medium"
    count: int = 0


class BatchAnalyzeItem(BaseModel):
    owner: str
    repo: str
    pr_number: int


class BatchAnalyzeRequest(BaseModel):
    prs: list[BatchAnalyzeItem]
    post_comment: bool = Field(False, description="分析完成后是否自动发布评论到 PR")


class BatchRiskCard(BaseModel):
    pr_number: int
    title: str
    risk_score: int
    risk_level: str


class BatchOverview(BaseModel):
    total_prs: int
    avg_risk_score: float
    highest_risk_pr: Optional[BatchRiskCard]
    risk_distribution: dict
    top_risks: list[str]
    risk_amplification: int = 0
    amplification_reason: str = ""


class FileOverlapItem(BaseModel):
    filename: str
    pr_numbers: list[int]
    changes_detail: list[dict]


class SimilarCodeBlockItem(BaseModel):
    block_hash: str
    pr_numbers: list[int]
    files: list[str]
    similarity_score: float
    snippet_preview: str


class DuplicateRiskPatternItem(BaseModel):
    description: str
    affected_prs: list[int]
    severity: str
    occurrence_count: int


class CrossPRDuplicateResult(BaseModel):
    file_overlaps: list[FileOverlapItem]
    similar_code_blocks: list[SimilarCodeBlockItem]
    duplicate_risk_patterns: list[DuplicateRiskPatternItem]
    summary: str


class BatchAnalyzeResponse(BaseModel):
    results: list[AnalyzeResponse]
    overview: BatchOverview
    duplicate_analysis: Optional[CrossPRDuplicateResult] = None


class CustomRuleCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=100)
    description: Optional[str] = Field(None, max_length=500)
    match_type: str = Field(..., pattern="^(text|regex|glob)$")
    match_pattern: str = Field(..., min_length=1)
    match_scope: str = Field("added_lines", pattern="^(added_lines|context_lines|full_file)$")
    file_filter: Optional[str] = Field(None)
    severity: str = Field("medium", pattern="^(critical|high|medium|low)$")
    suggestion: Optional[str] = Field(None, max_length=500)
    is_enabled: bool = True


class CustomRuleUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=100)
    description: Optional[str] = Field(None, max_length=500)
    match_type: Optional[str] = Field(None, pattern="^(text|regex|glob)$")
    match_pattern: Optional[str] = Field(None, min_length=1)
    match_scope: Optional[str] = Field(None, pattern="^(added_lines|context_lines|full_file)$")
    file_filter: Optional[str] = Field(None)
    severity: Optional[str] = Field(None, pattern="^(critical|high|medium|low)$")
    suggestion: Optional[str] = Field(None, max_length=500)
    is_enabled: Optional[bool] = None


class CustomRuleResponse(BaseModel):
    id: int
    name: str
    description: Optional[str] = None
    match_type: str
    match_pattern: str
    match_scope: str
    file_filter: Optional[str] = None
    severity: str
    suggestion: Optional[str] = None
    is_enabled: bool
    is_preset: bool
    created_at: Optional[str] = None
    updated_at: Optional[str] = None


class TrendDataPoint(BaseModel):
    day: Optional[str] = None
    pr_count: int = 0
    avg_risk_score: float = 0.0
    total_files_changed: int = 0
    total_additions: int = 0
    total_deletions: int = 0
    critical_count: int = 0
    high_count: int = 0
    medium_count: int = 0
    low_count: int = 0


class TrendSummary(BaseModel):
    total_prs: int = 0
    avg_risk_score: float = 0.0
    trend_direction: str = "stable"
    most_common_severity: str = "low"


class TrendResponse(BaseModel):
    data_points: list[TrendDataPoint] = []
    summary: TrendSummary


class FeedbackItem(BaseModel):
    index: int
    verdict: str = Field(..., pattern="^(accepted|false_positive|helpful|not_helpful)$")
    category: str = Field(..., pattern="^(risk_item|suggestion)$")


class FeedbackRequest(BaseModel):
    items: list[FeedbackItem]

File: src/types/review.ts
export interface AnalyzeRequest {
  owner: string
  repo: string
  prNumber: number
  postComment?: boolean
}

export interface PRFile {
  filename: string
  status: "modified" | "added" | "removed"
  additions: number
  deletions: number
  patch: string
}

export interface PRInfo {
  owner: string
  repo: string
  number: number
  title: string
  description: string
  author: string
  base_branch: string
  head_branch: string
  files_changed: number
  additions: number
  deletions: number
  files: PRFile[]
  diff_content: string
}

export type RiskSeverity = "critical" | "high" | "medium" | "low"

export interface RiskItem {
  severity: RiskSeverity
  file: string
  line: number
  description: string
  suggestion: string
  confidence?: number
  is_false_positive?: boolean
}

export type SuggestionCategory = "security" | "performance" | "maintainability" | "best-practice" | "bug-risk"

export interface SuggestionItem {
  category: SuggestionCategory
  description: string
  file: string
  code_snippet: string
  confidence?: number
}

export type RiskLevel = "low" | "medium" | "high" | "critical"

export interface RuleMatch {
  rule_id: number
  rule_name: string
  severity: string
  file: string
  line: number
  matched_text: string
  suggestion: string | null
}

export interface AnalyzeResponse {
  pr_info: PRInfo
  summary: string
  risk_items: RiskItem[]
  suggestions: SuggestionItem[]
  rule_matches: RuleMatch[]
  risk_score: number
  risk_level: RiskLevel
  estimated_review_minutes: number
  analysis_id?: number
}

export interface AnalyzeSuccessResponse {
  success: true
  data: AnalyzeResponse
}

export interface AnalyzeErrorResponse {
  success: false
  error: string
}

export type AnalyzeResult = AnalyzeSuccessResponse | AnalyzeErrorResponse

export interface HistoryItem {
  id: number
  repo_owner: string
  repo_name: string
  pr_number: number
  pr_title: string
  author: string
  risk_score: number
  risk_level: RiskLevel
  files_changed: number
  status: string
  created_at: string
}

export interface HistoryDetail extends HistoryItem {
  summary: string
  risk_items: RiskItem[]
  suggestions: SuggestionItem[]
  pr_description: string
  base_branch: string
  head_branch: string
  additions: number
  deletions: number
  estimated_review_minutes: number
  feedback?: Record<string, { verdict: string; timestamp?: string | null }> | null
  confidence_scores?: Record<string, number> | null
  diff_content?: string | null
}

export interface CustomRule {
  id: number
  name: string
  description: string | null
  match_type: "text" | "regex" | "glob"
  match_pattern: string
  match_scope: "added_lines" | "context_lines" | "full_file"
  file_filter: string | null
  severity: "critical" | "high" | "medium" | "low"
  suggestion: string | null
  is_enabled: boolean
  is_preset: boolean
  created_at: string | null
  updated_at: string | null
}

export interface TrendDataPoint {
  day: string | null
  pr_count: number
  avg_risk_score: number
  total_files_changed: number
  total_additions: number
  total_deletions: number
  critical_count: number
  high_count: number
  medium_count: number
  low_count: number
}

export interface TrendSummary {
  total_prs: number
  avg_risk_score: number
  trend_direction: "improving" | "worsening" | "stable"
  most_common_severity: string
}

export interface TrendResponse {
  data_points: TrendDataPoint[]
  summary: TrendSummary
}

export interface BatchAnalyzeItem {
  owner: string
  repo: string
  pr_number: number
}

export interface BatchOverview {
  total_prs: number
  avg_risk_score: number
  highest_risk_pr: {
    pr_number: number
    title: string
    risk_score: number
    risk_level: RiskLevel
  } | null
  risk_distribution: Record<string, number>
  top_risks: string[]
}

export interface BatchAnalyzeResponse {
  results: AnalyzeResponse[]
  overview: BatchOverview
  duplicate_analysis?: CrossPRDuplicateResult | null
}

export interface CrossPRDuplicateResult {
  file_overlaps: FileOverlapItem[]
  similar_code_blocks: SimilarCodeBlockItem[]
  duplicate_risk_patterns: DuplicateRiskPatternItem[]
  summary: string
}

export interface FileOverlapItem {
  filename: string
  pr_numbers: number[]
  changes_detail: { pr_number: number; status: string; additions: number; deletions: number }[]
}

export interface SimilarCodeBlockItem {
  block_hash: string
  pr_numbers: number[]
  files: string[]
  similarity_score: number
  snippet_preview: string
}

export interface DuplicateRiskPatternItem {
  description: string
  affected_prs: number[]
  severity: string
  occurrence_count: number
}

export interface FeedbackItem {
  index: number
  verdict: "accepted" | "false_positive" | "helpful" | "not_helpful"
  category: "risk_item" | "suggestion"
}

export interface FeedbackRequest {
  items: FeedbackItem[]
}

export const RISK_SEVERITY_CONFIG: Record<RiskSeverity, { label: string; color: string; bgClass: string }> = {
  critical: { label: "致命", color: "#ef4444", bgClass: "bg-red-500/10 text-red-400 border-red-500/30" },
  high: { label: "高危", color: "#f59e0b", bgClass: "bg-amber-500/10 text-amber-400 border-amber-500/30" },
  medium: { label: "中等", color: "#3b82f6", bgClass: "bg-blue-500/10 text-blue-400 border-blue-500/30" },
  low: { label: "建议", color: "#06d6a0", bgClass: "bg-teal-500/10 text-teal-400 border-teal-500/30" },
}

export const SUGGESTION_CATEGORY_CONFIG: Record<SuggestionCategory, { label: string; icon: string }> = {
  security: { label: "安全", icon: "Shield" },
  performance: { label: "性能", icon: "Zap" },
  maintainability: { label: "可维护", icon: "Lightbulb" },
  "best-practice": { label: "最佳实践", icon: "Lightbulb" },
  "bug-risk": { label: "缺陷风险", icon: "Bug" },
}

export const SEVERITY_ORDER: Record<RiskSeverity, number> = {
  critical: 0,
  high: 1,
  medium: 2,
  low: 3,
}

export const RISK_LEVEL_CONFIG: Record<RiskLevel, { label: string; textClass: string; bgClass: string }> = {
  low: { label: "低风险", textClass: "text-emerald-400", bgClass: "bg-emerald-500/10" },
  medium: { label: "中风险", textClass: "text-yellow-400", bgClass: "bg-yellow-500/10" },
  high: { label: "高风险", textClass: "text-orange-400", bgClass: "bg-orange-500/10" },
  critical: { label: "严重风险", textClass: "text-red-400", bgClass: "bg-red-500/10" },
}
